azurerm/resources.go

```diff
@@ -0,0 +1,135 @@
+package azurerm
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/cycloidio/terracognita/provider"
+	"github.com/cycloidio/terracognita/tag"
+)
+
+// ResourceType is the type used to define all the Resources
+// from the Provider
+type ResourceType int
+
+//go:generate enumer -type ResourceType -addprefix azurerm_ -transform snake -linecomment
+const (
+	ResourceGroup          ResourceType = iota
+	VirtualMachine         ResourceType = iota
+	VirtualNetwork         ResourceType = iota
+	Subnet                 ResourceType = iota
+	NetworkInterface       ResourceType = iota
+	NetworkSecurityGroup   ResourceType = iota
+	VirtualMachineScaleSet ResourceType = iota
+)
+
+type rtFn func(ctx context.Context, a *azurerm, resourceType string, tags []tag.Tag) ([]provider.Resource, error)
+
+var (
+	resources = map[ResourceType]rtFn{
+		ResourceGroup:          resourceGroup,
+		VirtualMachine:         virtualMachines,
+		VirtualNetwork:         cacheVirtualNetworks,
+		Subnet:                 subnets,
+		NetworkInterface:       networkInterfaces,
+		NetworkSecurityGroup:   networkSecurityGroups,
+		VirtualMachineScaleSet: virtualMachineScaleSets,
+	}
+)
+
+func resourceGroup(ctx context.Context, a *azurerm, resourceType string, tags []tag.Tag) ([]provider.Resource, error) {
+	resourceGroup := a.azurer.GetResourceGroup()
```

Comment:
we should check if `resourceGroup` is `nil`. It should not be the case, but it will avoid to panic a few lines later. Since the resource group seems to be highly used but not always, later we could have the need to check if it's nil or not.

Reply:
In the current context, what do you want to do if `resourceGroup == nil`? Return an `error`? Return `nil` silently?

Reply:
For now it's OK. It's early optimization. We will when we'll get this use case. :)